(TK-174) Query based on status level

rlinehan opened the pull request:
Add support for a `level` query param. When this query param is used, the
status function for the services are called with that level. If no level is
specified, defaults to `info`.

Also adds ringutils to help handle HTTP errors.

~~This is based off of #3 and will likely need to be rebased after that is in.~~ This PR has been rebased and should now be able to be merged.

rlinehan committed:
(TK-174) Add ability to specify status level in query param

Adds support for a `level` query param. When this query param is used, the
status function for the services are called with that level. If no level is
specified, defaults to `info`.

rlinehan committed:
(TK-174) Validate that level query param matches schema

Validate that the level in the query params matches a schema (that it is one
of :critical, :info, or :debug).

Add a ringutils namespace (basically copied out of other libraries) which
provides middleware for better error handling for HTTP responses.

rlinehan committed:
(maint) Handle non-user-data-related schema errors

Validate that the functions used for returning the status from a
service/services have correct inputs and outputs and, if not, return a
well-formatted HTTP error.

rlinehan committed:
(maint) Clean up and refactor service tests

Add `with-status-service` macro for bootstrapping the status service, its
dependencies, and any other desired services, using a common config.

rlinehan commented on src/puppetlabs/trapperkeeper/services/status/ringutils.clj:
```diff
@@ -0,0 +1,65 @@
+;; TODO: Much of this code is copied from other projects. It should
+;; probably be moved into a shared library or some such.
```
I copied this straight out of file sync service (which copied a lot of this out of puppet server). Perhaps we should add these to https://github.com/puppetlabs/ring-middleware?

KevinCorcoran replied:
Ha, yeah ... mess proliferation.  Although it probably needs some cleanup before landing in ring-middleware.

cprice404 replied:
Can we create a jira for this cleanup work?

rlinehan replied:
@cprice404 TK-214

rlinehan commented:
Updated/rebased and should now be ready to be reviewed and merged.

rlinehan committed:
(TK-174) Use handler/api rather than handler/site

handler/site from Compojure adds a lot of middleware that we don't need.
handler/api just adds middleware related to adding the query params into the
request map in a more useful way. This commit updates the routes to use
handler/api instead of handler/site, so we aren't using middleware that we
don't need.

Also, take advantage of the fact that handler/api uses `wrap-keyword-params`
and thus that params in `:params` are keywords (whereas params in
`:query-params` are not).

rlinehan committed:
(TK-174) Remove unused middleware from ringutils

rlinehan commented:
Added commits to use `handler/api` and to remove unused middleware.

KevinCorcoran commented on src/puppetlabs/trapperkeeper/services/status/ringutils.clj:
```diff
@@ -0,0 +1,47 @@
+;;; Public
+
+(defn wrap-request-data-errors
+  "A rignt middleware that catches slingshot errors of :type
```
s/rignt/ring

KevinCorcoran commented:
:shipit:

rlinehan committed:
(maint) Fix typo in docstring

cprice404 commented on src/puppetlabs/trapperkeeper/services/status/status_core.clj:
```diff
@@ -63,22 +82,28 @@
+      (compojure/context "/v1" []
+        (compojure/GET "/services" [:as {params :params}]
+          (let [level (get-status-detail-level params)
+                statuses (call-status-fns (deref status-fns-atom) level)]
```
This is a little bit of a nitpick, and this wasn't introduced in this PR... but...

Looking over this namespace, it seems to me like the fact that there is an atom involved in our management of the state of the service is an implementation detail that doesn't necessarily need to bleed through here?

I _think_ it should always be the case that the registration of the status callback functions for various services will happen during lifecycle init phases, and thus the atom will be effectively immutable by the time we want to build the routes... in which case it might be better to just pass a plain-old map in to the route building fns?  And to the `update` function at the top of this namespace?  Then the use of the atom would be restricted to the service code, and since it's kind of tightly coupled with the TK service context, that makes some sense to me and seems like it might be a bit cleaner.

If you decide that's worth doing, feel free to merge after that change goes in, you don't need to wait another day for me to look at it again.

If you decide it's not worth doing, I'm good with this being merged w/o that change.

rlinehan replied:
This makes a lot of sense, and I think it's worth doing. I'm not sure that this can be done for the `update` function (which gets called by `register-status!`) since it uses `swap!` to update the atom, but it could be unless changed to create a partial that gets used with `swap!` in the service code, or it could just be moved into the service code completely. In either of those changes the tests would probably need to change, but that's not a huge deal.

In any case, I think I'll open up a separate PR for these changes.

cprice404 commented on test/puppetlabs/trapperkeeper/services/status/status_service_test.clj:
```diff
@@ -11,57 +11,83 @@
-                     "status" "foo status 2"}}
-             body)))))
+    (testing "returns latest status for all services"
+      (let [req (http-client/get "http://localhost:8180/status/v1/services")
```
Maybe rename this var to `resp`?  I was confused momentarily because I thought you were reading the body of the request rather than the response.

rlinehan replied:
Good point, that would make a lot more sense. I'll change that throughout this namespace.

cprice404 commented on test/puppetlabs/trapperkeeper/services/status/status_service_test.clj:
```diff
@@ -11,57 +11,83 @@
+                       "status" "foo status 2 :info"}}
+               body))))
+    (testing "uses status level from query param"
+      (let [req (http-client/get "http://localhost:8180/status/v1/services?level=debug")
```
Same here.

cprice404 commented:
+1 aside from minor comments.

rlinehan committed:
(maint) Rename req to resp in tests

Throughout the service tests, `req` was being used as a parameter name to
refer to the response from an http request. This commit renames that
everywhere to the more accurate `resp`.

rlinehan commented:
I've added a PR changing `req` to `resp` per @cprice404's comments. I think I'll open a separate PR to  remove references to the atom from the core code.

KevinCorcoran commented:
@cprice404 I'm going to merge this so we can start some other work on top of it - it looks like all of your feedback has been addressed in one way or another.
